### DISCUSSION TITLE
Wayland implementation

### DISCUSSION THREAD

**Discussion by LDprg** [winapps-org/winapps#19] [Ideas]
Wayland with freerdp isnt that hard, we just need select between xfreerdp and the wlfreerdp command. This could happen automatically (based on env vars) or on a config field.

I would prefer to select automatically, as long as no config field is set.

Edit: fix typo

**Comment by TheStachelfisch** [comment]
wlfreerdp doesn't feature support for seamless windows. Only for full sessions.
Recently in the new beta release a new backend using SDL2 was announced which should have native wayland support and it supports RAIL (Remote Applications integrated locally). wlfreerdp will not have RAIL implemented.

I myself haven't tested yet, if sdl-freerdp works well with Wayland and using RAIL.
But it might be worth a shot.

Running SDL2 using Wayland: https://wiki.archlinux.org/title/wayland#SDL2
Relevant: https://github.com/FreeRDP/FreeRDP/issues/6573#issuecomment-1652380669
FreeRDP Release implementing SDL2: https://github.com/FreeRDP/FreeRDP/releases/tag/3.0.0-beta1

**Comment by LDprg** [comment]
Thx for the information.

We should probably shift using sdl-freerdp instead (even it is beta for now, but wlfreerdp makes no sense without RAIL). For me it seems like sdl-freerdp has the same interface as xfreerdp or wlfreerdp. However I didn't find an actual documentation to verify this.
Furthermore sdl-freerdp would be able to simplify the code for winapps, as we could use it for x11 and wayland.

**Reply by LDprg** [reply]
I am not sure if sdl-freerdp currently supports RAIL (see [here](https://github.com/FreeRDP/FreeRDP/milestone/28))
However there's a PR and RAILs is listed on the next milestones of freerdp.

**Reply by TheStachelfisch** [reply]
> I am not sure if sdl-freerdp currently supports RAIL (see [here](https://github.com/FreeRDP/FreeRDP/milestone/28))
> However there's a PR and RAILs is listed on the next milestones of freerdp. 

Hmm yeah you're right, I missed that. When I built freerdp and checked the help command of sdl-freerdp it showed, that the /app parameter was there.
But still good to know, that it's being worked on

**Comment by Shaarigan** [comment]
+1 from here since Nobara Linux just shipps with Wayland only

**Reply by LDprg** [reply]
@Shaarigan Are you sure? I did use nobara a  while back ago (the official image). As far as I know you can switch wayland and x11 using gdm or sddm.

**Reply by Shaarigan** [reply]
Thought I read that on their website but they however focus on Wayland cause of VRR and some pathches focus on Wayland. I anyways never tried using X11 cause of Waydroid 🤷

>

**Comment by LDprg** [comment]
@oskardotglobal Maybe we should consider a alternative solution for wayland users especially for the rewrite, until sdl-freerdp is ready. 

So everyone feel free to point out a possible tool for this purpose.

The tool has to fulfill following things:
- Wayland support
- Rails support or something similar (Application only streaming) 
- Controllable via Rust or shell
- Fast and Simple (or automated) setup
- Fast integration into winapps rewrite 
- Clipboard support

If you know a tool that only partially fits also feel free to comment.

**Reply by TheStachelfisch** [reply]
Unfortunately as it seems there is no other alternative than waiting for sdl-freerdp. Other than maybe closed-source tools like VMWare Unity or similar. Though I think resorting to closed-source tools isn't the way here

**Reply by LDprg** [reply]
How about other remote desktop protocols (VNC)?

**Reply by oskardotglobal** [reply]
I don't know if VNC supports this per-application style of sharing

**Reply by Shaarigan** [reply]
No they don't.

> On windows, there was no concept of separate GUI sessions for a long time, until recently via Remote Desktop, which is a hybrid solution. Which means windows VNC servers essentially mirror the contents of the primary monitor in their own memory, and send that virtual copy over VNC. It is thus sometimes desirable to be able to limit which chunk of a physical display to mirror - hence, the feature

On Linux this workes however opening an X11 remote session

**Reply by Shaarigan** [reply]
I tried to google for alternatives but it seems there isn't anything. Other projects that support RAIL are built on top of FreeRDP

**Reply by LDprg** [reply]
I have only one idea left. How about a modified spice client, would this work out?

**Reply by oskardotglobal** [reply]
Well we could use any protocol along with a custom client if we can just crop the stream to a window (although we'd have to send the window positions to the host somehow or determine them via the image)

Or we just write our own protocol / client + server (*at second though, no*)

**Reply by Shaarigan** [reply]
Maybe worth looking at Waydroid?
Waydroid is an Android Container running Apps on near native Linux performance and they also have a windowed mode one could use to have seamless windows as well - and they're running on wayland by default

**Comment by Shaarigan** [comment]
Do you think it might work under [xwayland](https://wayland.freedesktop.org/xserver.html)?

**Reply by LDprg** [reply]
Nope xwayland does not work at least for freerdp. RAIL seems to be very unstable with freerdp.

**Reply by oskardotglobal** [reply]
Coming back to this since I recently switched to Wayland and xfreerdp 3.15 works out of the box on Gnome 48

**Comment by LDprg** [comment]
Another option would to take a look at https://github.com/casualsnek/cassowary which apparently seems to work on wayland (as people reported in the issues).

**Reply by Shaarigan** [reply]
Following some gaming distro discussion, it seems that KDE as well as at least Fedora are in decision to drop X11 in the future so Wayland might become a must have

**Reply by Shaarigan** [reply]
Can't find any source thats talking about Cassowary working on Wayland session. They seem to use freeRDP as well so they should have the same issues as WinApps on Wayland?

**Reply by LDprg** [reply]
Strange why are people then reporting cassowary to work on wayland in the github issues?

**Reply by Shaarigan** [reply]
Quote from the dev
`AFAIK wlfreerdp does not support RAIL or Remote apps https://github.com/FreeRDP/FreeRDP/issues/5076 The wlfreerdp is there just for testing or while using full sessions. So if you want to use remote apps its recommended to use xfreerdp with Xwayland until freerdp wayland client has RAIL support !`
https://github.com/casualsnek/cassowary/issues/95

So either you're wrong and it works using xfreerdp and xwaylandbridge or the author is wrong and it causes trouble at some other point. I don't know and honestly have no time nor a propper configured system right now to test it, sorry 🤷

**Comment by eylenburg** [comment]
> Following some gaming distro discussion, it seems that KDE as well as at least Fedora are in decision to drop X11 in the future so Wayland might become a must have

KDE will not drop X11 but Wayland will be the default with Plasma 6 (release next month). Only Fedora will drop X11.




> Can't find any source thats talking about Cassowary working on Wayland session. They seem to use freeRDP as well so they should have the same issues as WinApps on Wayland?

That's true, it uses FreeRDP. The main differences ti Winapps seem to be that it has a GUI for settings and that it auto-suspends the Windows VM if not currently in use

**Reply by Shaarigan** [reply]
I said
`are in decision to drop X11 in the future`

You are right that only Fedora set it for now but time will show however 🤷

**Comment by ludgie** [comment]
Hello guys!
Hope you're doing great

While digging on WinApps and Cassowary, I've got the Docker image that you provide, to work with RDP through xfreerdp on wayland (hyprland) through xwayland

It seems to work well, though on hyprland I'm getting a new invisible window called "RemoteApp Marker Window", which I can close through the Alt-Tab switcher
Also, the shadows on tooltips of folders (in case of explorer.exe), have the hyprland shadow blur effect (it seems the compositor affects the rdp "instance"?)

Opening multiple windows seems to work as well

Note that I'm not using xwaylandvideobridge nor anything else, just xfreerdp connected to the docker image
Using archlinux, but I guess it could work on any distro

The behaviour (as far as I've tested only on explorer.exe) seems to be the same as per using on xorg (I use xfce)

**Reply by ludgie** [reply]
Seems to be working under labwc (stacking wlroots), but it's more buggy (a lot of latency when moving and resizing the windows, doesn't seem to be "performance" lag?)
The invisible app doesn't show up though like in hyprland, neither I can't seem to toggle the alt-tab switch from Windows11

![screenshotl2](https://github.com/winapps-org/winapps/assets/77110723/e59ce5f0-0418-4f84-8f65-8181c41d8abd)
![screenshotl1](https://github.com/winapps-org/winapps/assets/77110723/f59b68ce-cef6-4d41-ba85-7ba974e66958)

**Reply by ludgie** [reply]
It'd be interesting if someone could test it on GNOME or KDE wayland to see if the "latency" issue comes from labwc/wlroots or is from xfreerdp (?)

**Reply by ludgie** [reply]
Another note, the latency issue seems to only appear in specific apps (explorer.exe)
Windows Terminal doesn't have the issue, so it may be xfreerdp?

**Reply by ludgie** [reply]
Video showcasing the "latency" error (I don't know how to call it)
Another error that seems to happen is that some windows move by their own (in case of paint.net installer)

labwc/wlroots

https://github.com/winapps-org/winapps/assets/77110723/dcd1194e-ff02-4ff6-9252-6f93b5b1670d

**Reply by ludgie** [reply]
paint.net working, though due to its toolbar being multiwindow (like in GIMP years ago), when changing window focus the toolbar remains on top of everything (another bug?)
The clipboard doesn't seem to work with images, but with text it does
![screenshotl3](https://github.com/winapps-org/winapps/assets/77110723/d285d160-dcbc-424d-87c0-aa77c1628fa6)

**Reply by ludgie** [reply]
Notes from xorg:
- "Latency" error from explorer.exe window isn't present
- The bug of the paint.net installer window moving by its own is present on xorg
- The bug where paint.net toolbar is always focused is present too
So, these last two "bugs" aren't wayland exclusive

I would guess the image clipboard bug is probably from my distro (some missing package) rather than SPICE or RDP

Summary: it's usable on wayland depending which program you need to use, your luck may vary

**Reply by AkechiShiro** [reply]
@ludg1e Freerdp is coming with full support for Wayland, with SDL3 I believe, @LDprg do you confirm SDL3 brings Wayland support ?

**Reply by ludgie** [reply]
@AkechiShiro Yes, I'm aware, but currently sdlfreerdp (afaik) doesn't support RAILS (https://github.com/FreeRDP/FreeRDP/pull/9169)
As well as who knows when will SDL3 release

**Reply by AkechiShiro** [reply]
SDL3 was already released @ludg1e beta testing is ongoing with FreeRDP, RAILS support is still missing but is coming in the near future.

**Reply by oskardotglobal** [reply]
Hello, thanks for the awesome writeup. The "RemoteApp Marker Window" appears on all distros, that is intended behaviour. We will support wayland fully once RAILS is available for sdlfreerdp

**Comment by skabeche** [comment]
Hi, is there any update on this? Wayland is becoming more present now and it would need to be supported.
Thanks for your great work.

**Reply by AkechiShiro** [reply]
Hi @skabeche

We're looking into solutions, but at the moment, we are still investigating, I've been looking on this : https://forum.qubes-os.org/t/windows-support-in-qubes/4288

But I haven't had the time to check if that works on Wayland or not : https://github.com/elliotkillick/qvm-create-windows-qube

Also it seems to be working under Plasma 6 KDE (Wayland) at the moment

**Reply by KaKi87** [reply]
Hi,

> it seems to be working under Plasma 6 KDE (Wayland) at the moment

What are the installation instructions for Plasma 6 on KDE neon with Wayland ?

Thank you

**Reply by AkechiShiro** [reply]
Hello @KaKi87 

I haven't tested KDE Neon, the instruction on how to install, would be the same as for Ubuntu, as KDE Neon is just an Ubuntu shipping the latest KDE I believe.

**Reply by KaKi87** [reply]
Then what are the installation instructions for Plasma 6 on Ubuntu with Wayland ?

Thank you

**Reply by oskardotglobal** [reply]
They are the same for every version of Ubuntu. Check the README

**Reply by KaKi87** [reply]
So, there are no differences between installing on X11 and installing on Wayland, even though one of the packages in Step 2 contains "x11" in its name ?

**Reply by raffaelemancuso** [reply]
I tried  xfreerdp on a Wayland session (so running on XWayland I guess)but it's full of graphical glitches

**Reply by oskardotglobal** [reply]
FreeRDP is often full of graphical glitches, even under X11. I suggest you try it anyways, your mileage may vary. You can also try sdl-freerdp, but that's not feature complete.

**Comment by raffaelemancuso** [comment]
I have tried sdl-freerdp on Fedora 41 KDE Spin Wayland desktop session.

I can connect to the VM with sdl-freerdp, however, when I set it as the freerdp backend for winapps, winapps  doesn't connect to the VM properly (the VM shows  up but then it gets stuck)

**Comment by LDprg** [comment]
@oskardotglobal I did ping the freerdp maintainer again in the PR related to RAILs (which we need for the wayland implementation). (See [here](https://github.com/FreeRDP/FreeRDP/pull/9169)).

I am planning to get more active again (especially spending work on the rust rewrite). However it might take a month or two since I am currently still very busy.

**Reply by oskardotglobal** [reply]
Yeah, I should be able to work on #308 more soon too

**Comment by DrKoneko** [comment]
+1 for me, I also use Nobara and it doesn't support X11 in any capacity but this project seems amazing for exclusive programs I'd like to use.

**Comment by oskardotglobal** [comment]
This is something increasingly important. I've gone ahead and created #779 as a tracking issue.
